### PR TITLE
Add authentication guide and samples

### PR DESCRIPTION
**Pull request by ekoleda-codaio**
It's a lot of content, let me know if you think it should be broken up into smaller pages.

Staged:
- https://coda-packs-sdk--1604.com.readthedocs.build/en/1604/guides/advanced/authentication/
- https://coda-packs-sdk--1604.com.readthedocs.build/en/1604/samples/topic/authentication/

### CODE REVIEW

File: docs/guides/advanced/authentication.md [hunk 1]
```diff
@@ -4,7 +4,523 @@ title: Authentication
...
+```ts
+pack.setUserAuthentication({
+  // After approving access, the user should select which instance they want to
+  // connect to.
+  postSetup: [{
+    type: coda.PostSetupType.SetEndpoint,
+    name: "SelectEndpoint",
+    description: "Select the site to connect to:",
+    // Generate the list of endpoint options.
+    getOptionsFormula: coda.makeMetadataFormula(async function (context) {
```
Comment (jonathan-codaio) [inline docs/guides/advanced/authentication.md:469]: I don't think you should need `makeMetadataFormula` here or anywhere else anymore. Can you see if it works without it? If for some reason it needs it, we can fix it pretty easily.

File: docs/guides/advanced/authentication.md [hunk 2]
```diff
@@ -4,7 +4,523 @@ title: Authentication
...
+```
+
+The step's `getOptionsFormula` works like [dynamic autocomplete][autocomplete_dynamic], fetching the list of available endpoints and returning a display name and value for each. After the user enters their credentials they will be prompted to select one of the endpoints in a dialog.
+
+<img src="../../../images/auth_setendpoint.png" srcset="../../../images/auth_setendpoint_2x.png 2x" class="screenshot" alt="Selecting from a list of endpoints">
+
+[View Sample Code][sample_selected_endpoint]{ .md-button }
+
+!!! warning "Connection name generated twice"
+    When using this feature the [`getConnectionName` function](#setting-account-names) is run multiple times: once before the endpoint is selected and again after. Make sure your code works correctly in both cases. Before the endpoint is available you can return a generic name, which will get overwritten later.
```
Comment (jonathan-codaio) [inline docs/guides/advanced/authentication.md:491]: Ah interesting, I didn't realize this. Wonder if we can clean this up.

**Review by jonathan-codaio — COMMENTED** [review]
Terrific stuff here, such a valuable guide.

Reviewed the guide content, haven't gotten to the sample code just yet.

File: docs/guides/advanced/authentication.md [hunk 3]
```diff
@@ -4,7 +4,523 @@ title: Authentication
...
-<!-- TODO: Migrate the content to this repo. -->
-[See here][doc] for more information on how to configure authentication.
+One of the key use cases for Packs is integrating Coda with other apps and services, which often involves some form of authentication. When building a Pack you specify the type of authentication required by the API and Coda handles the credentials exchange, token storage, and account management.
+
+[View Sample Code][samples]{ .md-button }
+
+
+## Using a Pack with authentication
+
+Packs that use system-wide authentication (all users use the Pack makers's credentials) don't require any additional setup and can be used just like Packs without any authentication.
```
[the remark this answers is not in the document]
Comment (ekoleda-codaio) [reply docs/guides/advanced/authentication.md:14]: Good callout, done.

File: docs/guides/advanced/authentication.md [hunk 4]
```diff
@@ -4,7 +4,523 @@ title: Authentication
...
+
+Users can sign in to additional accounts and change their sharing settings from the **Settings** tab of the Pack's side panel. Accounts can be reused across docs, and  users can manage all of their connected accounts on the [Account settings][account_settings] page.
+
+
+### Selecting an account
+
+In the formula editor the account is shown as the first parameter to the formula, and in the other dialogs the account to use is displayed as a dropdown list.
+
+=== "In the formula editor"
+    <img src="../../../images/auth_formula.png" srcset="../../../images/auth_formula_2x.png 2x" class="screenshot" alt="Account selection in the formula editor">
```
[the remark this answers is not in the document]
Comment (ekoleda-codaio) [reply docs/guides/advanced/authentication.md:33]: 😄 I took that theme from the Coda API docs and ran with it!

File: docs/guides/advanced/authentication.md [hunk 5]
```diff
@@ -4,7 +4,523 @@ title: Authentication
...
+After adding the code, build a new version of your Pack and then navigate to the **Settings** tab. There you'll see an **Add system authentication** button you can use to set the credentials.
+
+<img src="../../../images/auth_system.png" srcset="../../../images/auth_system_2x.png 2x" class="screenshot" alt="Entering system credentials">
+
+The types of authentication supported, as well as the additional settings, are described in the sections below.
+
+
+## Authentication types
+
+Coda supports a fixed set of authentication types which cover the most common patterns that APIs use. It's not possible to implement a completely custom authentication scheme, as Coda alone has access to the user's credentials. If your API's authentication needs can't be met by any of these types please [contact support][support].
```
[the remark this answers is not in the document]
Comment (ekoleda-codaio) [reply docs/guides/advanced/authentication.md:81]: Ya, better to look there before contacting support. Done.

File: docs/guides/advanced/authentication.md [hunk 6]
```diff
@@ -4,7 +4,523 @@ title: Authentication
...
+
+    ```ts
+    pack.setUserAuthentication({
+      type: coda.AuthenticationType.HeaderBearerToken,
+    });
+    ```
+
+    [View Sample Code][sample_auth_header]{ .md-button }
+
+=== "Custom header"
```
[the remark this answers is not in the document]
Comment (ekoleda-codaio) [reply docs/guides/advanced/authentication.md:110]: I felt like it made more sense to go "Header, Header, Query" vs "Header, Query, Header".

File: docs/guides/advanced/authentication.md [hunk 7]
```diff
@@ -4,7 +4,523 @@ title: Authentication
...
+    1. Sending an `Authorization: Basic` header using the client ID and secret.
+
+    No configuration is required, Coda will try them both to see what works.
+
+<!-- TODO: Progressive OAuth -->
+
+
+## Requiring authentication
+
+If a Pack includes any form of user authentication then by default all of the building blocks within that Pack are assumed to require a connected account. You can set the `connectionRequirement` property on individual formulas or sync tables to make an account optional or not used.
```
[the remark this answers is not in the document]
Comment (ekoleda-codaio) [reply docs/guides/advanced/authentication.md:309]: To date in the documentation and examples I've been pretending like the non-builder pattern doesn't exist. I don't think we actually need to nuke it, but I think it's much easier to tell the story with only one pattern.

File: docs/guides/advanced/authentication.md [hunk 8]
```diff
@@ -4,7 +4,523 @@ title: Authentication
...
+  "site_url": "https://vandelay.example.com"
+}
+```
+
+You can automatically extract this value and use it as the endpoint URL by setting the `endpointKey` field of the authentication configuration.
+
+```ts
+pack.setUserAuthentication({
+  // ...
+  endpointKey: "site_url",
```
[the remark this answers is not in the document]
Comment (ekoleda-codaio) [reply docs/guides/advanced/authentication.md:449]: Great catch, since it only works on that type. Done.

File: docs/guides/advanced/authentication.md (hunk 1 above)
Comment (ekoleda-codaio) [reply docs/guides/advanced/authentication.md:469]: Ya, you do need it here. I have a TODO to file a bug about this.

File: docs/samples/topic/authentication.md [hunk 9]
```diff
@@ -11,27 +11,128 @@ Default authentication is the most common. Specify this if each user of your pac
 Use system authentication if you as the pack author will provide the necessary tokens to successfully make http requests within your pack. An example would be if your pack returns weather forecasts and the API involved requires an API key, but individual users need not provide their own API key. You as the pack author will register an API key and provide it to Coda, and Coda will apply it to all pack requests regardless of the user.
```
Comment (jonathan-codaio) [inline docs/samples/topic/authentication.md:11]: Nit, just noticed that the older code uses inconsistent capitalization for pack and http.

File: docs/guides/advanced/authentication.md (hunk 2 above)
Comment (ekoleda-codaio) [reply docs/guides/advanced/authentication.md:491]: That would be nice. I have a TODO to file a related bug, which is that if your `getOptionFormula` fails with an error, your partially complete account sticks around, in an unusable state (since it has no endpoint set).

File: docs/samples/topic/authentication.md [hunk 10]
```diff
@@ -52,5 +153,163 @@ pack.setUserAuthentication({
...
+  // After approving access, the user should select which instance they want to
+  // connect to.
+  requiresEndpointUrl: true,
+  endpointDomain: "atlassian.com",
+  postSetup: [{
+    type: coda.PostSetupType.SetEndpoint,
+    name: "SelectEndpoint",
+    description: "Select the site to connect to:",
+    // Determine the list of sites they have access to.
+    getOptionsFormula: coda.makeMetadataFormula(async function (context) {
```
Comment (jonathan-codaio) [inline docs/samples/topic/authentication.md:262]: Ditto here, let's remove `makeMetadataFormula` if possible, otherwise let's fix the underlying issue there.

File: docs/guides/advanced/authentication.md (hunk 1 above)
Comment (ekoleda-codaio) [reply docs/guides/advanced/authentication.md:470]: Filed: https://golinks.io/bug/20669

**Review by jonathan-codaio — APPROVED** [review]
LGTM, but probably worth fixing the capitalization of pack and http in the existing sample preamble per the most recent comment.

File: docs/guides/advanced/authentication.md (hunk 2 above)
Comment (ekoleda-codaio) [reply docs/guides/advanced/authentication.md:492]: Bug filed: https://golinks.io/bug/20672

**Comment by ekoleda-codaio** [comment]
> LGTM, but probably worth fixing the capitalization of pack and http in the existing sample preamble per the most recent comment.

Ya, I kind of forgot to update that preamble at all. Let me do a deeper pass in a follow-up CL, as I also think the content is a little too wordy for the side panel in the Web IDE.

File: docs/guides/advanced/authentication.md (hunk 1 above)
Comment (ekoleda-codaio) [reply docs/guides/advanced/authentication.md]: Now that this was fixed (#1605) I've removed it from here and the Jira sample.
